File: java/Samples/Biometrics/CS/SimpleFacesSampleCS/EnrollFromImage.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Windows.Forms;
using Neurotec.Biometrics;
using Neurotec.Biometrics.Client;
using Neurotec.Images;
using Neurotec.Licensing;

namespace Neurotec.Samples
{
	public partial class EnrollFromImage : UserControl
	{
		#region Public constructor

		public EnrollFromImage()
		{
			InitializeComponent();
		}

		#endregion

		#region Private fields

		private NBiometricClient _biometricClient;
		private NSubject _subject;
		private bool? _isSegmentationActivated;

		private float _defaultMaxRoll;
		private float _defaultMaxYaw;

		#endregion

		#region Public properties

		public NBiometricClient BiometricClient
		{
			get { return _biometricClient; }
			set
			{
				_biometricClient = value;
				_defaultMaxRoll = _biometricClient.FacesMaximalRoll;
				_defaultMaxYaw = _biometricClient.FacesMaximalYaw;
				cbRollAngle.SelectedItem = _defaultMaxRoll;
				cbYawAngle.SelectedItem = _defaultMaxYaw;
			}
		}

		#endregion

		#region Private methods

		private void SetBiometricClientParams()
		{
			if (!_isSegmentationActivated.HasValue)
				_isSegmentationActivated = NLicense.IsComponentActivated("Biometrics.FaceSegmentsDetection");

			_biometricClient.FacesDetectAllFeaturePoints = _isSegmentationActivated.Value;
			_biometricClient.FacesDetectBaseFeaturePoints = _isSegmentationActivated.Value;
			_biometricClient.FacesDetermineGender = _isSegmentationActivated.Value;
			_biometricClient.FacesDetermineAge = _isSegmentationActivated.Value;
			_biometricClient.FacesDetectProperties = _isSegmentationActivated.Value;
			_biometricClient.FacesRecognizeEmotion = _isSegmentationActivated.Value;
			_biometricClient.FacesRecognizeExpression = _isSegmentationActivated.Value;
		}

		private void ExtractTemplate()
		{
			if (_subject == null) return;
			SetBiometricClientParams();

			// Extract template
			_biometricClient.BeginCreateTemplate(_subject, OnExtractDone, null);
		}

		private void OnExtractDone(IAsyncResult r)
		{
			if (InvokeRequired)
			{
				BeginInvoke(new AsyncCallback(OnExtractDone), r);
			}
			else
			{
				try
				{
					NBiometricStatus status = _biometricClient.EndCreateTemplate(r);
					if (status == NBiometricStatus.Ok)
					{
						lblStatus.Text = @"Template extracted";
						btnSaveTemplate.Enabled = true;
					}
					else
					{
						lblStatus.Text = string.Format("Extraction failed: {0}", status);
						btnSaveTemplate.Enabled = false;
					}
				}
				catch (Exception ex)
				{
					Utils.ShowException(ex);
					lblStatus.Text = "Extraction failed!";
					btnSaveTemplate.Enabled = false;
				}
			}
		}

		#endregion

		#region Private form events

		private void EnrollFromImageLoad(object sender, EventArgs e)
		{
			try
			{
				float item = _biometricClient.FacesMaximalRoll;
				var items = new List<float>();
				for (float i = 15; i <= 180; i += 15)
				{
					items.Add((i));
				}
				if (!items.Contains(item))
					items.Add(item);
				items.Sort();

				int index = items.IndexOf(item);
				for (int i = 0; i != items.Count; i++)
				{
					cbRollAngle.Items.Add(items[i]);
				}
				cbRollAngle.SelectedIndex = index;

				item = _biometricClient.FacesMaximalYaw;
				items.Clear();
				for (float i = 15; i <= 90; i += 15)
				{
					items.Add((i));
				}
				if (!items.Contains(item))
					items.Add(item);
				items.Sort();

				index = items.IndexOf(item);
				for (int i = 0; i != items.Count; i++)
				{
					cbYawAngle.Items.Add(items[i]);
				}
				cbYawAngle.SelectedIndex = index;

				openFileDialog.Filter = NImages.GetOpenFileFilterString(true, true);
			}
			catch (Exception ex)
			{
				Utils.ShowException(ex);
			}
		}

		private void BtnSaveTemplateClick(object sender, EventArgs e)
		{
			if (saveFileDialog.ShowDialog() == DialogResult.OK)
			{
				try
				{
					File.WriteAllBytes(saveFileDialog.FileName, _subject.GetTemplateBuffer().ToArray());
				}
				catch (Exception ex)
				{
					Utils.ShowException(ex);
				}
			}
		}

		private void TsbExtractClick(object sender, EventArgs e)
		{
			try
			{
				ExtractTemplate();
			}
			catch (Exception ex)
			{
				Utils.ShowException(ex);
				btnSaveTemplate.Enabled = false;
			}
		}

		private void TsbOpenImageClick(object sender, EventArgs e)
		{
			try
			{
				openFileDialog.FileName = null;
				if (openFileDialog.ShowDialog() == DialogResult.OK)
				{
					// Create new subject
					_subject = new NSubject();
					var face = new NFace { FileName = openFileDialog.FileName };
					_subject.Faces.Add(face);
					facesView.Face = face;
					ExtractTemplate();
				}
			}
			catch (Exception ex)
			{
				Utils.ShowException(ex);
				btnSaveTemplate.Enabled = false;
			}
		}

		private void CbYawAngleSelectedIndexChanged(object sender, EventArgs e)
		{
			_biometricClient.FacesMaximalYaw = (float)cbYawAngle.SelectedItem;
		}

		private void CbRollAngleSelectedIndexChanged(object sender, EventArgs e)
		{
			_biometricClient.FacesMaximalRoll = (float)cbRollAngle.SelectedItem;
		}

		#endregion
	}
}

File: java/Samples/Biometrics/VB/SimpleFacesSampleVB/EnrollFromImage.vb
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Windows.Forms
Imports Neurotec.Biometrics
Imports Neurotec.Biometrics.Client
Imports Neurotec.Images
Imports Neurotec.Licensing

Partial Public Class EnrollFromImage
	Inherits UserControl
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()
	End Sub

#End Region

#Region "Private fields"

	Private _biometricClient As NBiometricClient
	Private _subject As NSubject
	Private _isSegmentationActivated As Boolean?

#End Region

#Region "Public properties"

	Public Property BiometricClient() As NBiometricClient
		Get
			Return _biometricClient
		End Get
		Set(ByVal value As NBiometricClient)
			_biometricClient = value
		End Set
	End Property

#End Region

#Region "Private methods"

	Private Sub SetBiometricClientParams()
		If Not _isSegmentationActivated.HasValue Then
			_isSegmentationActivated = NLicense.IsComponentActivated("Biometrics.FaceSegmentsDetection")
		End If

		_biometricClient.FacesDetectAllFeaturePoints = _isSegmentationActivated.Value
		_biometricClient.FacesDetectBaseFeaturePoints = _isSegmentationActivated.Value
		_biometricClient.FacesDetermineGender = _isSegmentationActivated.Value
		_biometricClient.FacesDetermineAge = _isSegmentationActivated.Value
		_biometricClient.FacesDetectProperties = _isSegmentationActivated.Value
		_biometricClient.FacesRecognizeEmotion = _isSegmentationActivated.Value
		_biometricClient.FacesRecognizeExpression = _isSegmentationActivated.Value
	End Sub

	Private Sub ExtractTemplate()
		If _subject Is Nothing Then
			Return
		End If
		SetBiometricClientParams()

		' Extract template
		_biometricClient.BeginCreateTemplate(_subject, AddressOf OnExtractDone, Nothing)
	End Sub

	Private Sub OnExtractDone(ByVal r As IAsyncResult)
		If InvokeRequired Then
			BeginInvoke(New AsyncCallback(AddressOf OnExtractDone), r)
		Else
			Try
				Dim status As NBiometricStatus = _biometricClient.EndCreateTemplate(r)
				If status = NBiometricStatus.Ok Then
					lblStatus.Text = "Template extracted"
					btnSaveTemplate.Enabled = True
				Else
					lblStatus.Text = String.Format("Extraction failed: {0}", status)
					btnSaveTemplate.Enabled = False
				End If
			Catch ex As Exception
				Utils.ShowException(ex)
				lblStatus.Text = "Extraction failed!"
				btnSaveTemplate.Enabled = False
			End Try
		End If
	End Sub

#End Region

#Region "Private form events"

	Private Sub EnrollFromImageLoad(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
		Try
			lblStatus.Text = String.Empty
			Dim item As Single = _biometricClient.FacesMaximalRoll
			Dim items As New List(Of Single)()
			For i As Single = 15 To 180 Step 15
				items.Add((i))
			Next i
			If (Not items.Contains(item)) Then
				items.Add(item)
			End If
			items.Sort()

			Dim index As Integer = items.IndexOf(item)
			Dim j As Integer = 0
			Do While j <> items.Count
				cbRollAngle.Items.Add(items(j))
				j += 1
			Loop
			cbRollAngle.SelectedIndex = index

			item = _biometricClient.FacesMaximalYaw
			items.Clear()
			For i As Single = 15 To 90 Step 15
				items.Add((i))
			Next i
			If (Not items.Contains(item)) Then
				items.Add(item)
			End If
			items.Sort()

			index = items.IndexOf(item)
			j = 0
			Do While j <> items.Count
				cbYawAngle.Items.Add(items(j))
				j += 1
			Loop
			cbYawAngle.SelectedIndex = index

			openFileDialog.Filter = NImages.GetOpenFileFilterString(True, True)
		Catch ex As Exception
			Utils.ShowException(ex)
		End Try
	End Sub

	Private Sub BtnSaveTemplateClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnSaveTemplate.Click
		If saveFileDialog.ShowDialog() = DialogResult.OK Then
			Try
				File.WriteAllBytes(saveFileDialog.FileName, _subject.GetTemplateBuffer().ToArray())
			Catch ex As Exception
				Utils.ShowException(ex)
			End Try
		End If
	End Sub

	Private Sub TsbExtractClick(ByVal sender As Object, ByVal e As EventArgs) Handles tsbExtract.Click
		Try
			ExtractTemplate()
		Catch ex As Exception
			Utils.ShowException(ex)
			btnSaveTemplate.Enabled = False
		End Try
	End Sub

	Private Sub TsbOpenImageClick(ByVal sender As Object, ByVal e As EventArgs) Handles tsbOpenImage.Click
		Try
			openFileDialog.FileName = Nothing
			If openFileDialog.ShowDialog() = DialogResult.OK Then
				' Create new subject
				_subject = New NSubject()
				Dim face As New NFace() With {.FileName = openFileDialog.FileName}
				_subject.Faces.Add(face)
				facesView.Face = face
				ExtractTemplate()
			End If
		Catch ex As Exception
			Utils.ShowException(ex)
			btnSaveTemplate.Enabled = False
		End Try
	End Sub

	Private Sub CbYawAngleSelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbYawAngle.SelectedIndexChanged
		_biometricClient.FacesMaximalYaw = CSng(cbYawAngle.SelectedItem)
	End Sub

	Private Sub CbRollAngleSelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbRollAngle.SelectedIndexChanged
		_biometricClient.FacesMaximalRoll = CSng(cbRollAngle.SelectedItem)
	End Sub

#End Region
End Class
